python/ql/lib/semmle/python/dataflow/new/internal/ImportResolution.qll

```diff
@@ -1,14 +1,72 @@
+/**
+ * INTERNAL. DO NOT USE.
+ *
+ * Provides predicates for resolving imports.
+ */
+
 private import python
 private import semmle.python.dataflow.new.DataFlow
 private import semmle.python.dataflow.new.internal.ImportStar
 private import semmle.python.dataflow.new.TypeTracker
 
+/**
+ * Python modules and the way imports are resolved are... complicated. Here's a crash course in how
+ * it works, as well as some caveats to bear in mind when looking at the implementation in this
+ * module.
+ *
+ * First, let's consider the humble `import` statement:
+ * ```python
+ * import foo
+ * import bar.baz
+ * import ham.eggs as spam
+ * ```
+ *
+ * In the AST, all imports are aliased, as in the last import above. That is, `import foo` becomes
+ * `import foo as foo`, and `import bar.baz` becomes `import bar as bar`. Note that `import` is
```

Comment:
`import bar.baz` must expand to more than just `import bar as bar`, perhaps also to `from bar import baz as baz`?

Reply:
No. The guarantee that `import bar.baz` gives is that subsequent uses of `bar.baz` seen during execution will resolve to that module. (Technically, it also has an effect on the order of initialisation, but this is irrelevant to our modelling of module resolution.)

Yes, this is confusing. 🙂